core/lib/Thelia/Action/Product.php

```diff
@@ -100,6 +121,451 @@ public function create(ProductCreateEvent $event)
+                $this->setClonePSEAssociatedDocuments($cp->getId(), $cpPSEId, $opPSEDocuments);
+            }
+        }
+
+        $this->setCloneFeatureCombination($opId, $cp->getId(), $dispatcher);
+
+        $this->setCloneProductAssociatedContent($opId, $cp, $dispatcher);
+
+        $this->setCloneI18n($lang, $op, $cp, $dispatcher);
+    }
+
+    public function createClone(ProductModel $op, $opDefaultI18n, $opDefaultPrice, ProductCloneEvent $event)
+    {
+        // Build event and dispatch creation of the clone product
+        $createCloneEvent = new ProductCreateEvent();
+        $createCloneEvent
+            ->setTitle($opDefaultI18n->getTitle())
+            ->setRef($event->getRef())
+            ->setLocale($event->getLang())
+            ->setVisible(0)
+            ->setVirtual($op->getVirtual())
+            ->setTaxRuleId($op->getTaxRuleId())
+            ->setDefaultCategory($op->getDefaultCategoryId())
+            ->setBasePrice($opDefaultPrice->getPrice())
+            ->setCurrencyId($opDefaultPrice->getCurrencyId())
+            ->setBaseWeight($op->getDefaultSaleElements()->getWeight());
+
+        $event->getDispatcher()->dispatch(TheliaEvents::PRODUCT_CREATE, $createCloneEvent);
+
+        $event->setCpId($createCloneEvent->getProduct()->getId());
+
+        return $createCloneEvent->getProduct();
+    }
+
+    public function updateClone(ProductModel $cp, ProductModel $op, $opDefaultI18n, $opDefaultPrice, EventDispatcherInterface $dispatcher)
+    {
+        // Set other product's information
+        $cpUpdateEvent = new ProductUpdateEvent($cp->getId());
+        $cpUpdateEvent
+            ->setRef($cp->getRef())
+            ->setVisible($cp->getVisible())
+            ->setVirtual($cp->getVirtual())
+            ->setTitle($opDefaultI18n->getTitle())
+            ->setLocale($opDefaultI18n->getLocale())
+
+            ->setBasePrice($opDefaultPrice->getPrice())
+            ->setBaseWeight($op->getDefaultSaleElements()->getWeight())
+            ->setTaxRuleId($op->getTaxRuleId())
+            ->setCurrencyId($opDefaultPrice->getCurrencyId())
+
+            ->setChapo($opDefaultI18n->getChapo())
+            ->setDescription($opDefaultI18n->getDescription())
+            ->setPostscriptum($opDefaultI18n->getPostscriptum())
+            ->setBrandId($op->getBrandId())
+
+            ->setDefaultCategory($op->getDefaultCategoryId());
+
+        $dispatcher->dispatch(TheliaEvents::PRODUCT_UPDATE, $cpUpdateEvent);
+
+        $cp = $cpUpdateEvent->getProduct();
+
+        // Set clone's template
+        $cpUpdateTemplateEvent = new ProductSetTemplateEvent($cp, $op->getTemplateId(), $opDefaultPrice->getCurrencyId());
+        $dispatcher->dispatch(TheliaEvents::PRODUCT_SET_TEMPLATE, $cpUpdateTemplateEvent);
+
+        return $cp;
+    }
+
+    public function setCloneFiles($opId, ProductModel $cp, EventDispatcherInterface $dispatcher)
+    {
```

Comment:
This method should not be in the Product action. You have to create a cloning method in BaseCachedFile, and the related event.